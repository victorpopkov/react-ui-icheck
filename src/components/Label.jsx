import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Label extends Component {
  constructor(props) {
    super(props);

    this.handleHover = this.handleHover.bind(this);
  }

  handleHover(event, isHovered = true) {
    const { disabled, handleHover, isMobile, labelHover } = this.props;

    if (disabled) return;

    if (handleHover && labelHover && !isHovered) {
      handleHover(event, false);
    } else if (handleHover) {
      handleHover(event, isHovered);
    }

    if (isMobile) event.stopPropagation();
  }

  renderLabel() {
    const { label, labelChildren, labelTag, labelTagClassName } = this.props;

    if (!React.isValidElement(label)) {
      const Tag = labelTag;

      if (labelChildren) {
        return <Tag className={labelTagClassName}>{labelChildren}</Tag>;
      }

      return (
        // eslint-disable-next-line react/no-danger
        <Tag
          className={labelTagClassName}
          dangerouslySetInnerHTML={{ __html: label }}
        />
      );
    }

    return null;
  }

  render() {
    const { children, labelClassName } = this.props;

    return (
      <label
        className={labelClassName || null}
        onBlur={(event) => this.handleHover(event, false)}
        onFocus={this.handleHover}
        onMouseOut={(event) => this.handleHover(event, false)}
        onMouseOver={this.handleHover}
        onTouchEnd={(event) => this.handleHover(event, false)}
        onTouchStart={this.handleHover}
      >
        {children}
        {this.renderLabel()}
      </label>
    );
  }
}

Label.propTypes = {
  children: PropTypes.node.isRequired,
  handleHover: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  isMobile: PropTypes.bool,
  label: PropTypes.node,
  labelChildren: PropTypes.node,
  labelClassName: PropTypes.string,
  labelHover: PropTypes.bool,
  labelTag: PropTypes.string,
  labelTagClassName: PropTypes.string,
};

Label.defaultProps = {
  disabled: false,
  isMobile: false,
  label: null,
  labelChildren: null,
  labelClassName: null,
  labelHover: true,
  labelTag: 'span',
  labelTagClassName: null,
};

export default Label;
