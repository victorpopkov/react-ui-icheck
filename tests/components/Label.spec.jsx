import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import expect from 'expect';
import { fireEvent, render } from '@testing-library/react';
import Label from '../../src/components/Label';

describe('<Label />', () => {
  describe('when required props are not passed', () => {
    it('should call console.error()', () => {
      const consoleErrorSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});
      render(<Label />);
      expect(consoleErrorSpy).toHaveBeenCalledWith(
        'Warning: Failed %s type: %s%s',
        'prop',
        'The prop `children` is marked as required in `Label`, but its value is `undefined`.',
        expect.anything(),
      );
      consoleErrorSpy.mockRestore();
    });
  });

  describe('when required props are passed', () => {
    const testEventHandleHover = (description, event, called, arg) => {
      describe(`${description}`, () => {
        it(`${
          called > 0 ? 'should call' : 'should not call'
        } the \`handleHover\` prop function`, () => {
          const handleHover = jest.fn();
          const { container } = render(
            <Label handleHover={handleHover}>
              <div>test</div>
            </Label>,
          );
          expect(handleHover).toHaveBeenCalledTimes(0);
          fireEvent[event](container.querySelector('label'));
          expect(handleHover).toHaveBeenCalledTimes(called);
          expect(handleHover).toHaveBeenCalledWith(expect.anything(), arg);
        });
      });
    };

    it('should match the snapshot', () => {
      const { container } = render(
        <Label handleHover={jest.fn()}>
          <div>test</div>
        </Label>,
      );
      expect(container).toMatchSnapshot();
    });

    testEventHandleHover('and focused', 'focus', 1, true);
    testEventHandleHover('and blurred', 'blur', 1, false);
    testEventHandleHover('and has mouse over', 'mouseOver', 1, true);
    testEventHandleHover('and has mouse out', 'mouseOut', 1, false);
    testEventHandleHover('and touch starts', 'touchStart', 1, true);
    testEventHandleHover('and touch ends', 'touchEnd', 1, false);

    describe('when `label` prop is "test"', () => {
      it('should insert <span>test</span> after children', () => {
        const { container } = render(
          <Label handleHover={jest.fn()} label="test">
            <div>test</div>
          </Label>,
        );
        const labelElement = container.querySelector('div + span');
        expect(labelElement.tagName).toBe('SPAN');
        expect(labelElement.innerHTML).toEqual('test');
      });

      describe('and `labelChildren` prop is <div>test</div>', () => {
        it('should insert <span><div>test</div></span> after children', () => {
          const { container } = render(
            <Label
              handleHover={jest.fn()}
              label="test"
              labelChildren={<div>test</div>}
            >
              <div>test</div>
            </Label>,
          );
          const labelElement = container.querySelector('div + span');
          expect(labelElement.tagName).toBe('SPAN');
          expect(labelElement.innerHTML).toEqual('<div>test</div>');
        });
      });

      describe('and `labelTag` prop is "div"', () => {
        it('should insert <div>test</div> after children', () => {
          const { container } = render(
            <Label handleHover={jest.fn()} label="test" labelTag="div">
              <div>test</div>
            </Label>,
          );
          const labelElement = container.querySelector('div + div');
          expect(labelElement.tagName).toBe('DIV');
          expect(labelElement.innerHTML).toEqual('test');
        });
      });

      describe('and `labelTagClassName` prop is "test"', () => {
        it('should insert <span class="test">test</span> after children', () => {
          const { container } = render(
            <Label
              handleHover={jest.fn()}
              label="test"
              labelTagClassName="test"
            >
              <div>test</div>
            </Label>,
          );
          const labelElement = container.querySelector('div + span');
          expect(labelElement.tagName).toBe('SPAN');
          expect(labelElement.innerHTML).toEqual('test');
          expect(labelElement).toHaveClass('test');
        });
      });
    });
  });
});
