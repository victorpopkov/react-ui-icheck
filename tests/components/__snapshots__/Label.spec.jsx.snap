// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Label /> when required props are passed should match the snapshot 1`] = `
<div>
  <label>
    <div>
      test
    </div>
    <span />
  </label>
</div>
`;
